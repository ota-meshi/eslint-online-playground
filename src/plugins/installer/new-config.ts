import type * as ESTree from "estree";
import type {
  BuildESLintConfigHelper,
  ConfigInstallPluginResult,
  Plugin,
} from "..";
import { alertAndLog } from "./error";
import {
  analyzeScope,
  isDirective,
  isExportDefault,
  isModuleExports,
  isRequireDeclaration,
  iterateRequireDeclarator,
} from "../../utils/estree-utils";

export async function installPluginForFlatConfig(
  configText: string,
  plugins: Plugin[],
): Promise<ConfigInstallPluginResult> {
  const codeRead = await import("code-red");
  try {
    const ast: ESTree.Program = codeRead.parse(configText, {
      ecmaVersion: "latest",
      ranges: true,
      locations: true,
      sourceType: "module",
      allowReturnOutsideFunction: true,
    }) as never;

    let targetNode: ESTree.ArrayExpression;
    let sourceType: "module" | "script" = "module";
    const exportDefault = ast.body.find(isExportDefault);
    if (exportDefault) {
      sourceType = "module";
      if (exportDefault.declaration.type === "ArrayExpression") {
        targetNode = exportDefault.declaration;
      } else {
        targetNode = exportDefault.declaration = {
          type: "ArrayExpression",
          elements: [
            {
              type: "SpreadElement",
              argument: exportDefault.declaration as any,
            },
          ],
        };
      }
    } else {
      const commonJsExports = ast.body.find(isModuleExports);
      if (commonJsExports) {
        sourceType = "script";
        if (commonJsExports.expression.right.type === "ArrayExpression") {
          targetNode = commonJsExports.expression.right;
        } else {
          targetNode = commonJsExports.expression.right = {
            type: "ArrayExpression",
            elements: [
              {
                type: "SpreadElement",
                argument: commonJsExports.expression.right,
              },
            ],
          };
        }
      } else {
        alertAndLog(
          "Failed to parse eslint.config.*. Failed to add new configuration.",
        );
        return { error: true };
      }
    }

    const scopeManager = await analyzeScope(ast);
    const usedIds = new Set<string>();
    scopeManager.globalScope.through.forEach((ref) =>
      usedIds.add(ref.identifier.name),
    );
    scopeManager.globalScope.variables.forEach((variable) =>
      usedIds.add(variable.name),
    );
    scopeManager.globalScope.childScopes
      .find((s) => s.type === "module")
      ?.variables.forEach((variable) => usedIds.add(variable.name));

    const helper: BuildESLintConfigHelper = {
      type: sourceType,
      x(code) {
        return codeRead.parseExpressionAt(code, 0, {});
      },
      i(code) {
        const decl = codeRead.parse(code, {
          ecmaVersion: "latest",
          sourceType: "module",
        }).body[0];

        if (decl?.type === "ImportDeclaration") {
          return decl;
        }
        throw new Error("Failed to parse import declaration.");
      },
      spread(expression) {
        return {
          type: "SpreadElement",
          argument: expression,
        };
      },
    };

    for (const plugin of plugins) {
      if (!plugin.eslintConfig) {
        alertAndLog("Contains plugins that do not support new configurations.");
        return { error: true };
      }
      const eslintConfig = plugin.eslintConfig;
      const localNames: Record<string, string> = Object.create(null);
      for (const decl of eslintConfig.imports(helper)) {
        if (sourceType === "module") {
          margeImport(ast.body, decl, localNames, usedIds);
        } else {
          margeRequire(ast.body, decl, localNames, usedIds);
        }
      }
      targetNode.elements.push(...eslintConfig.expression(localNames, helper));
    }

    return { configText: codeRead.print(ast).code };
  } catch (e) {
    // eslint-disable-next-line no-console -- ignore
    console.error(e);
    alertAndLog("Failed to parse config. Failed to add new configuration.");
    return { error: true };
  }
}

function margeImport(
  body: (ESTree.Directive | ESTree.Statement | ESTree.ModuleDeclaration)[],
  decl: ESTree.ImportDeclaration,
  localNames: Record<string, string>,
  usedIds: Set<string>,
) {
  for (const spec of decl.specifiers) {
    if (spec.type === "ImportNamespaceSpecifier") {
      const node = findImportAll();
      if (node) {
        localNames[spec.local.name] = node.spec.local.name;
      } else {
        const name = resolveName(spec.local.name, usedIds);
        insertImport({
          type: "ImportDeclaration",
          source: decl.source,
          specifiers: [
            {
              type: "ImportNamespaceSpecifier",
              local: {
                type: "Identifier",
                name,
              },
            },
          ],
        });
        localNames[spec.local.name] = name;
        usedIds.add(name);
      }
    } else if (spec.type === "ImportDefaultSpecifier") {
      const node = findImportDefault();
      if (node) {
        localNames[spec.local.name] = node.spec.local.name;
      } else {
        const name = resolveName(spec.local.name, usedIds);
        insertImport({
          type: "ImportDeclaration",
          source: decl.source,
          specifiers: [
            {
              type: "ImportDefaultSpecifier",
              local: {
                type: "Identifier",
                name,
              },
            },
          ],
        });
        localNames[spec.local.name] = name;
        usedIds.add(name);
      }
    } else {
      const node = findNamedImport(spec.imported.name);
      if (node) {
        localNames[spec.local.name] = node.spec.local.name;
      } else {
        const name = resolveName(spec.local.name, usedIds);
        insertImport({
          type: "ImportDeclaration",
          source: decl.source,
          specifiers: [
            {
              type: "ImportSpecifier",
              imported: {
                type: "Identifier",
                name: spec.imported.name,
              },
              local: {
                type: "Identifier",
                name,
              },
            },
          ],
        });
        localNames[spec.local.name] = name;
        usedIds.add(name);
      }
    }
  }

  function findImportAll():
    | { decl: ESTree.ImportDeclaration; spec: ESTree.ImportNamespaceSpecifier }
    | undefined {
    for (const node of iterateImport()) {
      const ns = node.specifiers.find(
        (spec): spec is ESTree.ImportNamespaceSpecifier =>
          spec.type === "ImportNamespaceSpecifier",
      );
      if (ns) {
        return { decl: node, spec: ns };
      }
    }
    return undefined;
  }

  function findImportDefault():
    | { decl: ESTree.ImportDeclaration; spec: ESTree.ImportDefaultSpecifier }
    | undefined {
    for (const node of iterateImport()) {
      const d = node.specifiers.find(
        (spec): spec is ESTree.ImportDefaultSpecifier =>
          spec.type === "ImportDefaultSpecifier",
      );
      if (d) {
        return { decl: node, spec: d };
      }
    }
    return undefined;
  }

  function findNamedImport(
    imported: string,
  ):
    | { decl: ESTree.ImportDeclaration; spec: ESTree.ImportSpecifier }
    | undefined {
    for (const node of iterateImport()) {
      const spec = node.specifiers.find(
        (spec): spec is ESTree.ImportSpecifier =>
          spec.type === "ImportSpecifier" && spec.imported.name === imported,
      );
      if (spec) {
        return { decl: node, spec };
      }
    }
    return undefined;
  }

  function* iterateImport(): Iterable<ESTree.ImportDeclaration> {
    for (const node of body) {
      if (
        node.type === "ImportDeclaration" &&
        node.source.value === decl.source.value
      )
        yield node;
    }
  }

  function insertImport(node: ESTree.ImportDeclaration) {
    for (let index = 0; index < body.length; index++) {
      const st = body[index];
      if (st.type === "ImportDeclaration" || isDirective(st)) {
        continue;
      }
      body.splice(index, 0, node);
      return;
    }
    body.push(node);
  }
}

function margeRequire(
  body: (ESTree.Directive | ESTree.Statement | ESTree.ModuleDeclaration)[],
  decl: ESTree.ImportDeclaration,
  localNames: Record<string, string>,
  usedIds: Set<string>,
) {
  let requireNode = findRequire();
  if (!requireNode) {
    requireNode = insertRequireFromSource();
  }
  for (const spec of decl.specifiers) {
    let importedName: string | null;
    if (spec.type === "ImportNamespaceSpecifier") {
      importedName = null;
    } else if (spec.type === "ImportDefaultSpecifier") {
      importedName = "default";
    } else {
      importedName = spec.imported.name;
    }
    const preferName: string = spec.local.name;
    if (requireNode.type === "Identifier") {
      localNames[preferName] = `${requireNode.name}${
        importedName ? `.${importedName}` : ""
      }`;
    } else if (importedName) {
      const prop = requireNode.properties.find(
        (
          prop,
        ): prop is ESTree.AssignmentProperty & {
          value: ESTree.Identifier;
        } =>
          prop.type === "Property" &&
          !prop.computed &&
          prop.key.type === "Identifier" &&
          prop.key.name === importedName &&
          prop.value.type === "Identifier",
      );
      if (prop) {
        localNames[preferName] = prop.value.name;
      } else {
        const localName = resolveName(preferName, usedIds);
        usedIds.add(localName);
        requireNode.properties.push({
          type: "Property",
          computed: false,
          shorthand: false,
          method: false,
          kind: "init",
          key: {
            type: "Identifier",
            name: importedName,
          },
          value: {
            type: "Identifier",
            name: localName,
          },
        });
        localNames[preferName] = localName;
      }
    } else {
      requireNode = insertRequireFromSource();
      localNames[preferName] = `${requireNode.name}${
        importedName ? `.${importedName}` : ""
      }`;
    }
  }

  function findRequire() {
    for (const node of body) {
      for (const v of iterateRequireDeclarator(node)) {
        const moduleName = v.init.arguments[0];
        if (
          moduleName?.type === "Literal" &&
          moduleName.value === decl.source.value &&
          (v.id.type === "Identifier" || v.id.type === "ObjectPattern")
        ) {
          return v.id;
        }
      }
    }
    return null;
  }

  function insertRequireFromSource() {
    const localName = resolveName(
      toId(String(decl.source.value).replace(/^eslint-/u, "")),
      usedIds,
    );
    usedIds.add(localName);
    const id: ESTree.Identifier = {
      type: "Identifier",
      name: localName,
    };
    insertRequire({
      type: "VariableDeclaration",
      kind: "const",
      declarations: [
        {
          type: "VariableDeclarator",
          id,
          init: {
            type: "CallExpression",
            callee: {
              type: "Identifier",
              name: "require",
            },
            arguments: [
              {
                type: "Literal",
                value: String(decl.source.value),
              },
            ],
            optional: false,
          },
        },
      ],
    });
    return id;
  }

  function insertRequire(
    node: ESTree.VariableDeclaration & {
      declarations: [
        {
          type: "VariableDeclarator";
          id: ESTree.Identifier;
          init: ESTree.SimpleCallExpression;
        },
      ];
    },
  ): void {
    for (let index = 0; index < body.length; index++) {
      const st = body[index];
      if (isRequireDeclaration(st) || isDirective(st)) {
        continue;
      }
      body.splice(index, 0, node);
      return;
    }
    body.push(node);
  }

  function toId(name: string) {
    return camelCase(name.replace(/\W/giu, "_"));
  }

  function camelCase(str: string) {
    return str
      .replace(/(?:\b|[-_])(\w)/gu, (_, c) => (c ? c.toUpperCase() : ""))
      .replace(/^./u, (c) => c.toLowerCase());
  }
}

function resolveName(name: string, usedIds: Set<string>) {
  let resolved = name;
  while (usedIds.has(resolved)) {
    resolved = `_${resolved}`;
  }
  return resolved;
}
