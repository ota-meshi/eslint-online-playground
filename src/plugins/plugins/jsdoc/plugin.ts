import type { ESLintConfig, ESLintLegacyConfig } from "../..";

export const name = "eslint-plugin-jsdoc";
export const description = "JSDoc linting rules for ESLint";
export const repo = "https://github.com/gajus/eslint-plugin-jsdoc";
export const devDependencies = { [name]: "latest" };
export const eslintLegacyConfig: ESLintLegacyConfig = {
  extends: ["plugin:jsdoc/recommended"],
};
export const eslintConfig: ESLintConfig<"jsdoc"> = {
  *imports(helper) {
    if (helper.type === "module") {
      yield helper.i(`import jsdoc from 'eslint-plugin-jsdoc'`);
    } else {
      yield helper.i(`import * as jsdoc from 'eslint-plugin-jsdoc'`);
    }
  },
  *expression(names, helper) {
    yield helper.x(`${names.jsdoc}.configs['flat/recommended']`);
  },
};
