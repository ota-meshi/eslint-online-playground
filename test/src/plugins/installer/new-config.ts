import { expect, describe, test } from "vitest";

import { installPluginForFlatConfig } from "../../../../src/plugins/installer/new-config.ts";

const esmConfigText = `import js from "@eslint/js";
import jsdoc from 'eslint-plugin-jsdoc';
export default [
  js.configs.recommended,
  {
    rules: {
      quotes: ["error", "double"],
    },
  },
];
`;
const cjsConfigText = `const js = require("@eslint/js");
const jsdoc  = require('eslint-plugin-jsdoc');
module.exports =  [
  js.configs.recommended,
  {
    rules: {
      quotes: ["error", "double"],
    },
  },
];
`;

describe("installPluginForFlatConfig with es-x", async () => {
  const plugin = await import("../../../../src/plugins/plugins/es-x/plugin.ts");
  test("esm", async () => {
    const result = await installPluginForFlatConfig(esmConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
  test("cjs", async () => {
    const result = await installPluginForFlatConfig(cjsConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
});
describe("installPluginForFlatConfig with jsdoc", async () => {
  const plugin = await import(
    "../../../../src/plugins/plugins/jsdoc/plugin.ts"
  );
  test("esm", async () => {
    const result = await installPluginForFlatConfig(esmConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
  test("cjs", async () => {
    const result = await installPluginForFlatConfig(cjsConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
});
describe("installPluginForFlatConfig with stylistic", async () => {
  const plugin = await import(
    "../../../../src/plugins/plugins/stylistic/plugin.ts"
  );
  test("esm", async () => {
    const result = await installPluginForFlatConfig(esmConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
  test("cjs", async () => {
    const result = await installPluginForFlatConfig(cjsConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
});
describe("installPluginForFlatConfig with typescript-eslint", async () => {
  const plugin = await import(
    "../../../../src/plugins/plugins/typescript-eslint/plugin.ts"
  );
  test("esm", async () => {
    const result = await installPluginForFlatConfig(esmConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
  test("cjs", async () => {
    const result = await installPluginForFlatConfig(cjsConfigText, [plugin]);
    expect(result.configText).toMatchSnapshot();
  });
});
