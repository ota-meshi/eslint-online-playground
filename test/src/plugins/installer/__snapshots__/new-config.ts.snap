// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`installPluginForFlatConfig with es-x > cjs 1`] = `
"const js = require("@eslint/js");
const jsdoc = require('eslint-plugin-jsdoc');
const es = require('eslint-plugin-es-x');

module.exports = [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	es.configs['flat/restrict-to-es2019']
];"
`;

exports[`installPluginForFlatConfig with es-x > esm 1`] = `
"import js from "@eslint/js";
import jsdoc from 'eslint-plugin-jsdoc';
import es from 'eslint-plugin-es-x';

export default [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	es.configs['flat/restrict-to-es2019']
];"
`;

exports[`installPluginForFlatConfig with jsdoc > cjs 1`] = `
"const js = require("@eslint/js");
const jsdoc = require('eslint-plugin-jsdoc');

module.exports = [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	jsdoc.configs['flat/recommended']
];"
`;

exports[`installPluginForFlatConfig with jsdoc > esm 1`] = `
"import js from "@eslint/js";
import jsdoc from 'eslint-plugin-jsdoc';

export default [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	jsdoc.configs['flat/recommended']
];"
`;

exports[`installPluginForFlatConfig with stylistic > cjs 1`] = `
"const js = require("@eslint/js");
const jsdoc = require('eslint-plugin-jsdoc');
const stylistic = require('@stylistic/eslint-plugin');

module.exports = [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	stylistic.configs['recommended-flat']
];"
`;

exports[`installPluginForFlatConfig with stylistic > esm 1`] = `
"import js from "@eslint/js";
import jsdoc from 'eslint-plugin-jsdoc';
import stylistic from '@stylistic/eslint-plugin';

export default [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	stylistic.configs['recommended-flat']
];"
`;

exports[`installPluginForFlatConfig with typescript-eslint > cjs 1`] = `
"const js = require("@eslint/js");
const jsdoc = require('eslint-plugin-jsdoc');
const tseslint = require('typescript-eslint');

module.exports = [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	...tseslint.config(...tseslint.configs.recommended)
];"
`;

exports[`installPluginForFlatConfig with typescript-eslint > esm 1`] = `
"import js from "@eslint/js";
import jsdoc from 'eslint-plugin-jsdoc';
import tseslint from 'typescript-eslint';

export default [
	js.configs.recommended,
	{ rules: { quotes: ["error", "double"] } },
	...tseslint.config(...tseslint.configs.recommended)
];"
`;
